### PR TITLE
Canonicalize `std.dim` into `flow.dispatch.shape` when relevant.

### PR DESCRIPTION
**Pull request by silvasean**
Linalg tends to create `std.dim` ops because it is upstream code.
However, inside a `flow.dispatch.workgroups` region, we should be using
`flow.dispatch.shape` to get the shape of the dispatch tensors.

This plays nicer with the way that flow.dispatch.workgroups is meant to
be used, and results in much nicer IR as it lets everything fold away.

Example IR before/after this change, as seen by the
LinalgLLVMBufferizePass
https://gist.github.com/silvasean/60b80d442cde72ca4a39f6198a7075dc

The new code results in no dim ops at the LinalgLLVMBufferizePass, so we
can consider removing support for them there. I don't do that in this
patch since I'm not sure if there are other considerations in flight.
I believe that by running dim op canonicalization to a fixed point we
can guarantee that they always go away by construction before that pass.

### CODE REVIEW

**Review by benvanik — APPROVED** [review]
nice!

**Comment by MaheshRavishankar** [comment]
UhOh, I dont know what all changed here, but I was making some conflicting changes to this part. Basically need to have the `hal.interface.binding.subspan` have a `tensor/memref` return type instead of a flow type and move all the shape computation to MaterializeInterface pass. Its a bit non-trivial to me how these all intersect.

**Comment by MaheshRavishankar** [comment]
I guess I will need to rebase on top of these after these lands. Thanks for handling this Sean. I might ping you if something breaks after i rebase (waiting on integrates to land)

**Comment by silvasean** [comment]
Sorry Mahesh! Feel free to grab me if you need any help.

Generally speaking, the should mainly affect things at the flow level, before `hal.interface.binding.subspan` is introduced. Specifically, OutlineDispatchRegions2Pass outlines and inserts the `flow.dispatch.tie_shape` ops that allow all the `flow.dispatch.shape` ops introduced here to resolve and disappear.

**Pull request merged**
